server/src/main/java/org/jboss/as/server/suspend/SuspendController.java

```diff
@@ -5,274 +5,323 @@
+ *  </ol>
+ * </li>
+ * <li>Set state to {@code State#SUSPENDED}</li>
+ * </ol>
+ * Resuming the suspended server happens in one phase:
+ * <ol>
+ * <li>Iterate over activity groups in reverse priority order (from last to first). For each group:
+ *  <ol>
+ *  <li>Create resume stages for each registered server activity</li>
+ *  <li>Once all resume stages within priority group have complete, move on to next priority group</li>
+ *  </ol>
+ * </li>
+ * <li>Set state to {@code State#RUNNING}</li>
+ * </ol>
  * @author Stuart Douglas
+ * @author Paul Ferraro
  */
-public class SuspendController implements Service<SuspendController> {
+public class SuspendController implements ServerSuspendController, SuspendableActivityRegistry {
+    private static final Supplier<List<SuspendableActivity>> FACTORY = CopyOnWriteArrayList::new;
 
-    /**
-     * Timer that handles the timeout. We create it on pause, rather than leaving it hanging round.
-     */
-    private Timer timer;
-
-    private State state = State.SUSPENDED;
-
-    private final NavigableMap<Integer, List<ServerActivity>> activitiesByGroup = new TreeMap<>();
-
-    private final List<OperationListener> operationListeners = new ArrayList<>();
-
-    private final InjectedValue<NotificationHandlerRegistry> notificationHandlerRegistry = new InjectedValue<>();
+    // Server activity groups in suspend priority order (max -> min)
+    // Initialized as an unmodifiable list of empty activity lists
+    private final List<List<SuspendableActivity>> activityGroups = Stream.generate(FACTORY).limit(SuspendPriority.LAST.ordinal() + 1).collect(Collectors.toUnmodifiableList());
+    // Index of activity priorities
+    private final Map<SuspendableActivity, SuspendPriority> priorities = Collections.synchronizedMap(new IdentityHashMap<>());
 
-    private int groupsCount;
+    private final List<OperationListener> listeners = new CopyOnWriteArrayList<>();
 
-    private boolean startSuspended;
+    private volatile State state = State.SUSPENDED;
 
-    private final ServerActivityCallback listener = this::activityPaused;
-
-    public SuspendController() {
-        this.startSuspended = false;
-    }
-
-    public void setStartSuspended(boolean startSuspended) {
-        //TODO: it is not very clear what this boolean stands for now.
-        this.startSuspended = startSuspended;
-        state = State.SUSPENDED;
+    @Override
+    public void reset() {
+        this.state = State.SUSPENDED;
     }
 
-    public synchronized void suspend(long timeoutMillis) {
-        if(state == State.SUSPENDED) {
-            return;
-        }
-        if (timeoutMillis > 0) {
-            ServerLogger.ROOT_LOGGER.suspendingServer(timeoutMillis);
-        } else if (timeoutMillis < 0) {
-            ServerLogger.ROOT_LOGGER.suspendingServerWithNoTimeout();
-        } else {
-            ServerLogger.ROOT_LOGGER.suspendingServer();
```

Comment:
They were not removed (though it might be hard to tell via the diff).
See: https://github.com/pferraro/wildfly-core/blob/WFCORE-6894/server/src/main/java/org/jboss/as/server/suspend/SuspendController.java#L220

Reply:
@pferraro  Sorry, I did not explain it well. I mean, they are no longer shown, because now we are using the new methods, and the usage of the new methods do not take into account these traces, for example now the timeout is used via `CompletableFuture#completeOnTimeout` so it is completely detached.

Reply:
Somehow, these traces should be invoked via `suspend(ServerSuspendContext context)`, or by the external handlers that invoke the suspend, since they know which kind of timeout is bein used

Reply:
I see what you mean.  It would be the responsibility of the caller of suspend(context) to log whether or not it intends to wait for the CompletionStage of the suspend to finish.  I will make sure those log messages happen where it is appropriate.

Reply:
Fixed.